RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitPhase2WeightsProducerGPU.cc

```diff
@@ -0,0 +1,117 @@
+
+  uint32_t n_;
+};
+
+// constructor with initialisation of elements
+EcalUncalibRecHitPhase2WeightsProducerGPU::EcalUncalibRecHitPhase2WeightsProducerGPU(const edm::ParameterSet &ps)
+    : weights_(ps.getParameter<std::vector<double>>("weights")),
+      digisToken_{consumes<InputProduct>(ps.getParameter<edm::InputTag>("digisLabelEB"))},
+      recHitsToken_{produces<OutputProduct>(ps.getParameter<std::string>("recHitsLabelEB"))} {}
+
+void EcalUncalibRecHitPhase2WeightsProducerGPU::fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<std::string>("recHitsLabelEB", "EcalUncalibRecHitsEB");
+  desc.add<std::vector<double>>("weights",
+                                {-0.121016,
+                                 -0.119899,
+                                 -0.120923,
+                                 -0.0848959,
+                                 0.261041,
+                                 0.509881,
+                                 0.373591,
+                                 0.134899,
+                                 -0.0233605,
+                                 -0.0913195,
+                                 -0.112452,
+                                 -0.118596,
+                                 -0.121737,
+                                 -0.121737,
+                                 -0.121737,
+                                 -0.121737});
+
+  desc.add<edm::InputTag>("digisLabelEB", edm::InputTag("simEcalUnsuppressedDigis", ""));
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+// aquire function which initislises objects on host and device to their actual objects and calls kernal
+void EcalUncalibRecHitPhase2WeightsProducerGPU::acquire(edm::Event const &event,
+                                                        edm::EventSetup const &setup,
+                                                        edm::WaitingTaskWithArenaHolder holder) {
+  // cuda products
+  auto const &digisProduct = event.get(digisToken_);
+  // raii
+  cms::cuda::ScopedContextAcquire ctx{digisProduct, std::move(holder), cudaState_};
+
+  // get actual obj
+  auto const &digis = ctx.get(digisProduct);
+
+  n_ = digis.size;
+
+  // if no digis stop here
+  if (n_ == 0)
+    return;
+
+  // weights to GPU
+
+  cms::cuda::device::unique_ptr<double[]> weights_d =
+      cms::cuda::make_device_unique<double[]>(EcalDataFrame_Ph2::MAXSAMPLES, ctx.stream());
+
+  cudaCheck(cudaMemcpyAsync(weights_d.get(),
+                            weights_.data(),
+                            EcalDataFrame_Ph2::MAXSAMPLES * sizeof(double),
+                            cudaMemcpyHostToDevice,
+                            ctx.stream()));
+
+  // output on GPU
+  eventOutputDataGPU_.allocate(n_, ctx.stream());
+
+  ecal::weights::entryPoint(digis, eventOutputDataGPU_, weights_d, ctx.stream());
+}
+
+void EcalUncalibRecHitPhase2WeightsProducerGPU::produce(edm::Event &event, const edm::EventSetup &setup) {
```

Comment:
I believe this EDProducer could work without `ExternalWork` as well.

Reply:
This EDProducer's `produce()` was essentially copied from the Phase 1 multifit GPU reconstruction where `ExternalWork` was used as well (https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitProducerGPU.cc#L271). Why would it be necessary there but not here?

Reply:
It seems to me the `EcalUncalibRecHitProducerGPU` would not need `ExternalWork` either. Looking quickly the history, originally the module had an option to transfer some data back to host (although in a bit weird way), that was removed at one point, so maybe the use of `ExternalWork` is a historic artefact.

If you prefer to keep the structure of these two modules similar, this simplification could be done in a subsequent PR for both.

Reply:
OK I see. I would prefer to do the simplification in a separate PR for both in that case.

Reply:
OK.
Could you open a separate issue to keep track of this future simplification ?